internal/api/hooks.go

```diff
@@ -55,20 +74,219 @@ func (a *API) runHook(ctx context.Context, tx *storage.Connection, name string,
+		req.Header.Set("Content-Type", "application/json")
+
+		req.Header.Set("webhook-id", msgID.String())
+		req.Header.Set("webhook-timestamp", fmt.Sprintf("%d", currentTime.Unix()))
+		req.Header.Set("webhook-signature", strings.Join(signatureList, ", "))
+
+		watcher, req := watchForConnection(req)
+		rsp, err := client.Do(req)
+
+		if err != nil {
+
+			if terr, ok := err.(net.Error); ok && terr.Timeout() {
+				hookLog.Errorf("Request timed out for attempt %d with err %s", i, err)
+				// TODO: workshop the sleep time value
+				time.Sleep(DefaultHTTPHookTimeout)
+				continue
+			} else if !watcher.gotConn && i < DefaultHTTPHookRetries-1 {
+				hookLog.Errorf("Failed to establish a connection on attempt %d with err %s", i, err)
+				time.Sleep(DefaultHTTPHookTimeout)
+				continue
+			} else if i == DefaultHTTPHookRetries-1 {
+				return nil, httpError(http.StatusGatewayTimeout, "Failed to perform webhook in time frame (%v seconds)", time.Since(startTime).Seconds())
+
+			} else {
+				return nil, internalServerError("Failed to trigger auth hook, error making HTTP request").WithInternalError(err)
+			}
+		}
+
+		switch rsp.StatusCode {
+		case http.StatusOK, http.StatusNoContent, http.StatusAccepted:
+			if rsp.Body == nil {
+				return nil, nil
+			}
+			defer rsp.Body.Close()
+			body, err := io.ReadAll(rsp.Body)
+			if err != nil {
+				return nil, err
+			}
+			if isOverSizeLimit(body) {
+				return nil, internalServerError("payload too large")
+			}
+			return body, nil
+		case http.StatusTooManyRequests, http.StatusServiceUnavailable:
+			retryAfterHeader := rsp.Header.Get("retry-after")
+			if retryAfterHeader != "true" {
+				continue
+			}
+			// TODO: maybe return errors properly for all of these from response
+			return []byte{}, tooManyRequestsError("too many requests")
+		case http.StatusBadRequest:
+			return nil, badRequestError("bad request error")
+		case http.StatusUnauthorized:
+			return []byte{}, unauthorizedError("unauthorized")
+		default:
+			return []byte{}, internalServerError("error executing hooks")
+		}
+	}
+	return nil, internalServerError("error executing hook")
+}
+
+func generateSignatures(secrets []string, msgID uuid.UUID, currentTime time.Time, inputPayload []byte) ([]string, error) {
+	// TODO(joel): Handle asymmetric case once library has been upgraded
+	var signatureList []string
+	for _, secret := range secrets {
+		if strings.HasPrefix(secret, SymmetricSignaturePrefix) {
+			trimmedSecret := strings.TrimPrefix(secret, SymmetricSignaturePrefix)
+			wh, err := standardwebhooks.NewWebhook(trimmedSecret)
+			if err != nil {
+				return nil, err
+			}
+			signature, err := wh.Sign(msgID.String(), currentTime, inputPayload)
+			if err != nil {
+				return nil, err
+			}
+			signatureList = append(signatureList, signature)
+		} else {
+			return nil, errors.New("invalid signature format")
+		}
+	}
+	return signatureList, nil
+}
+
+func watchForConnection(req *http.Request) (*connectionWatcher, *http.Request) {
```

Comment:
Why is tjis? Can't we just use normal HTTP functions?

Reply:
The intention is to check if a connection has been established before sending over a payload (e.g. DNS resolution connection pool exhaustion). This helps distinguish between a scenario where a connection was made and there was an error (internal server error) as opposed to the case where the connection could not be established as all

It was from the original package and though it was useful. More details [here](https://pkg.go.dev/net/http/httptrace#GotConnInfo)

Is there a way to perform similar functionality with the HTTP package?

Reply:
IMO I think it's not that necessary for now.

Reply:
Okay will remove it then